### PR TITLE
CNS 720: providers rewards pool research and design

### PR DESCRIPTION
**Pull request by Yaroms**

### CODE REVIEW

File: x/rewards/keeper/providers.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+			if totalRewarded.GT(total) {
+				utils.LavaFormatError("trying to send more than we can", nil, utils.LogAttr("total", total.String()), utils.LogAttr("totalRewarded", totalRewarded.String()))
+				return
+			}
+			// now give the reward the provider contributor and delegators
+			providerAddr, err := sdk.AccAddressFromBech32(basepay.Provider)
+			if err != nil {
+				continue
+			}
+			_, err = k.dualstakingKeeper.RewardProvidersAndDelegators(ctx, providerAddr, basepay.ChainID, reward, "reward pool", false, false, false)
```
Comment (Yaroms) [inline x/rewards/keeper/providers.go:43]: change to the real pool name

File: x/rewards/keeper/providers_test.go [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+
+	res2, err := ts.QueryDualstakingDelegatorRewards(providerAcc2.Addr.String(), "", "")
+	require.Nil(t, err)
+	require.Len(t, res.Rewards, 1)
+	require.Equal(t, res2.Rewards[0].Amount.Amount, distBalance.QuoRaw(7).MulRaw(2))
+
+	res3, err := ts.QueryDualstakingDelegatorRewards(providerAcc3.Addr.String(), "", "")
+	require.Nil(t, err)
+	require.Len(t, res.Rewards, 1)
+	require.Equal(t, res3.Rewards[0].Amount.Amount, distBalance.QuoRaw(7).MulRaw(4).AddRaw(6))
```
Comment (oren-lava) [inline x/rewards/keeper/providers_test.go:381]: add a comment the add 6 and sub 3 are because of remainder issues

**Review by oren-lava — CHANGES_REQUESTED** [review]

File: x/rewards/keeper/providers.go (hunk 1 above)
Comment (Yaroms) [reply x/rewards/keeper/providers.go]: done

File: proto/lavanet/lava/rewards/params.proto [hunk 3]
```diff
@@ -32,4 +32,10 @@ message Params {
     (gogoproto.customtype) = "github.com/cosmos/cosmos-sdk/types.Dec",
     (gogoproto.nullable)   = false
   ];
+
+  Providers providers =5 [(gogoproto.nullable)=false];
+}
+
+message Providers {
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply proto/lavanet/lava/rewards/params.proto:39]: done

File: x/rewards/keeper/base_pay.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+	if b == nil {
+		return val, false
+	}
+
+	k.cdc.MustUnmarshal(b, &val)
+	return val, true
+}
+
+// RemoveBasePay removes a BasePay from the store
+func (k Keeper) removeBasePay(
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/rewards/keeper/base_pay.go:33]: done

File: x/rewards/keeper/base_pay.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+	ctx sdk.Context,
+	index types.BasePayIndex,
+) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefix(types.BasePayPrefix))
+	store.Delete([]byte(index.String()))
+}
+
+// GetAllBasePay returns all BasePay
+func (k Keeper) getAllBasePay(ctx sdk.Context) (list []types.BasePayWithIndex) {
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/rewards/keeper/base_pay.go:42]: done

File: x/rewards/keeper/providers.go [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+		}
+	}
+	k.removeAllBasePay(ctx)
+	tokensToBurn := total.Sub(totalRewarded)
+	k.BurnPoolTokens(ctx, types.ProviderDistributionPool, tokensToBurn)
+}
+
+func (k Keeper) SpecTotalPayout(ctx sdk.Context, totalMonthlyPayout math.Int, totalProvidersBaseRewards sdk.Dec, spec types.SpecEmmisionPart) math.LegacyDec {
+	specPayoutAllocation := spec.Emission.MulInt(totalMonthlyPayout)
+	// TODO yarom move the param read to a method
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/rewards/keeper/providers.go:63]: done

File: x/rewards/keeper/providers.go [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+			}
+		}
+	}
+	k.removeAllBasePay(ctx)
+	err := k.bankKeeper.SendCoinsFromModuleToModule(ctx, string(types.ProviderDistributionPool), subscriptionTypes.ModuleName, sdk.NewCoins(sdk.NewCoin(k.stakingKeeper.BondDenom(ctx), totalRewarded)))
+	if err != nil {
+		utils.LavaFormatError("Failed to send bonus rewards to subscription module", err, utils.LogAttr("amount", totalRewarded.String()))
+	}
+	tokensToBurn := total.Sub(totalRewarded)
+	k.BurnPoolTokens(ctx, types.ProviderDistributionPool, tokensToBurn)
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/rewards/keeper/providers.go:58]: done

File: x/rewards/keeper/providers.go [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+func (k Keeper) DistributeMonthlyBonusRewards(ctx sdk.Context) {
+	total := k.TotalPoolTokens(ctx, types.ProviderDistributionPool)
+	totalRewarded := sdk.ZeroInt()
+	specs := k.SpecEmissionParts(ctx)
+	for _, spec := range specs {
+		basepays, totalbasepay := k.SpecProvidersBasePay(ctx, spec.ChainID)
+		specTotalPayout := k.SpecTotalPayout(ctx, total, sdk.NewDecFromInt(totalbasepay), spec)
+		for _, basepay := range basepays {
+			reward := specTotalPayout.Mul(basepay.TotalAdjusted).QuoInt(totalbasepay).TruncateInt()
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/rewards/keeper/providers.go:34]: done

File: x/rewards/keeper/providers.go [hunk 9]
```diff
@@ -0,0 +1,101 @@
+package keeper
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/rewards/keeper/providers.go:1]: done

File: x/rewards/keeper/providers.go [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+	totalRewarded := sdk.ZeroInt()
+	specs := k.SpecEmissionParts(ctx)
+	for _, spec := range specs {
+		basepays, totalbasepay := k.SpecProvidersBasePay(ctx, spec.ChainID)
+		specTotalPayout := k.SpecTotalPayout(ctx, total, sdk.NewDecFromInt(totalbasepay), spec)
+		for _, basepay := range basepays {
+			reward := specTotalPayout.Mul(basepay.TotalAdjusted).QuoInt(totalbasepay).TruncateInt()
+			totalRewarded = totalRewarded.Add(reward)
+			if totalRewarded.GT(total) {
+				utils.LavaFormatError("trying to send more than we can", nil, utils.LogAttr("total", total.String()), utils.LogAttr("totalRewarded", totalRewarded.String()))
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/rewards/keeper/providers.go:37]: done

File: x/rewards/keeper/providers.go [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+		totalStake = totalStake.Add(chainStake[chainID])
+	}
+
+	for _, chainID := range chainIDs {
+		if stake, ok := chainStake[chainID]; ok {
+			if stake.IsZero() {
+				continue
+			}
+
+			emisions = append(emisions, types.SpecEmmisionPart{ChainID: chainID, Emission: stake.Quo(totalStake)})
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/rewards/keeper/providers.go:93]: done

File: x/rewards/types/base_pay.go [hunk 12]
```diff
@@ -0,0 +1,36 @@
+package types
+
+import (
+	"encoding/binary"
+	fmt "fmt"
+)
+
+var _ binary.ByteOrder
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/rewards/types/base_pay.go:8]: deleted

File: x/rewards/genesis.go [hunk 13]
```diff
@@ -12,6 +12,7 @@ import (
 func InitGenesis(ctx sdk.Context, k keeper.Keeper, genState types.GenesisState) {
 	// this line is used by starport scaffolding # genesis/module/init
 	k.SetParams(ctx, genState.Params)
+	// TODO yarom: add base pay and timerstore
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/rewards/genesis.go:15]: done

File: x/rewards/genesis.go [hunk 14]
```diff
@@ -20,6 +21,9 @@ func InitGenesis(ctx sdk.Context, k keeper.Keeper, genState types.GenesisState)
 func ExportGenesis(ctx sdk.Context, k keeper.Keeper) *types.GenesisState {
 	genesis := types.DefaultGenesis()
 	genesis.Params = k.GetParams(ctx)
+
+	// TODO yarom: add base pay and timerstore
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/rewards/genesis.go:25]: done

File: x/rewards/keeper/providers_test.go (hunk 2 above)
Comment (Yaroms) [reply x/rewards/keeper/providers_test.go:381]: done

File: x/rewards/keeper/providers_test.go [hunk 15]
```diff
@@ -0,0 +1,382 @@
...
+
+	// first months there are no bonus rewards, just payment ffrom the subscription
+	ts.AdvanceMonths(1)
+	ts.AdvanceBlocks(ts.BlocksToSave() + 1)
+
+	res, err := ts.QueryDualstakingDelegatorRewards(providerAAcc.Addr.String(), providerAAcc.Addr.String(), "")
+	require.Nil(t, err)
+	require.Len(t, res.Rewards, 0)
+
+	// now the provider should get all of the provider allocation
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/rewards/keeper/providers_test.go:35]: done

File: x/rewards/keeper/providers_test.go [hunk 16]
```diff
@@ -0,0 +1,382 @@
...
+	"github.com/lavanet/lava/x/rewards/types"
+	subscription "github.com/lavanet/lava/x/subscription/keeper"
+	"github.com/stretchr/testify/require"
+)
+
+// the rewards here zero
+func TestZeroProvidersRewards(t *testing.T) {
+	ts := newTester(t)
+
+	providerAAcc, _ := ts.AddAccount(common.PROVIDER, 1, testBalance)
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/rewards/keeper/providers_test.go:17]: done

File: x/rewards/keeper/providers_test.go [hunk 17]
```diff
@@ -0,0 +1,382 @@
...
+	err := ts.StakeProvider(providerAAcc.Addr.String(), ts.spec, testBalance)
+	require.Nil(t, err)
+
+	ts.AdvanceEpoch()
+
+	consumerAAcc, _ := ts.AddAccount(common.CONSUMER, 1, ts.plan.Price.Amount.Int64())
+	_, err = ts.TxSubscriptionBuy(consumerAAcc.Addr.String(), consumerAAcc.Addr.String(), ts.plan.Index, 1, false)
+	require.Nil(t, err)
+
+	// first months there are no bonus rewards, just payment ffrom the subscription
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/rewards/keeper/providers_test.go:27]: done

File: x/rewards/keeper/providers_test.go [hunk 18]
```diff
@@ -0,0 +1,382 @@
...
+	ts.AdvanceMonths(1)
+	ts.AdvanceBlocks(ts.BlocksToSave() + 1)
+
+	res, err := ts.QueryDualstakingDelegatorRewards(providerAAcc.Addr.String(), providerAAcc.Addr.String(), "")
+	require.Nil(t, err)
+	require.Len(t, res.Rewards, 1)
+	//	require.Equal(t, res.Rewards[0].Amount, ts.plan.Price)
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/rewards/keeper/providers_test.go:108]: nope, done

File: x/rewards/keeper/providers_test.go [hunk 19]
```diff
@@ -0,0 +1,382 @@
...
+	require.Nil(t, err)
+
+	// first months there are no bonus rewards, just payment ffrom the subscription
+	ts.AdvanceMonths(1)
+	ts.AdvanceBlocks(ts.BlocksToSave() + 1)
+
+	res, err := ts.QueryDualstakingDelegatorRewards(providerAcc1.Addr.String(), "", "")
+	require.Nil(t, err)
+	require.Len(t, res.Rewards, 1)
+	ts.TxDualstakingClaimRewards(providerAcc1.Addr.String(), "")
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/rewards/keeper/providers_test.go:351]: done,
i dont need to check the rewards after, the claiming is done by another module

File: x/rewards/keeper/providers_test.go [hunk 20]
```diff
@@ -0,0 +1,382 @@
...
+
+	res, err := ts.QueryDualstakingDelegatorRewards(providerAAcc.Addr.String(), providerAAcc.Addr.String(), "")
+	require.Nil(t, err)
+	require.Len(t, res.Rewards, 0)
+
+	// now the provider should get all of the provider allocation
+	ts.AdvanceMonths(1)
+	ts.AdvanceEpoch()
+
+	res, err = ts.QueryDualstakingDelegatorRewards(providerAAcc.Addr.String(), providerAAcc.Addr.String(), "")
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/rewards/keeper/providers_test.go:39]: done

File: x/rewards/keeper/providers_test.go [hunk 21]
```diff
@@ -0,0 +1,382 @@
...
+	// first months there are no bonus rewards, just payment ffrom the subscription
+	ts.AdvanceMonths(1)
+	ts.AdvanceBlocks(ts.BlocksToSave() + 1)
+
+	res, err := ts.QueryDualstakingDelegatorRewards(providerAAcc.Addr.String(), providerAAcc.Addr.String(), "")
+	require.Nil(t, err)
+	require.Len(t, res.Rewards, 2)
+	require.Equal(t, res.Rewards[0].Amount, ts.plan.Price)
+	require.Equal(t, res.Rewards[1].Amount, ts.plan.Price)
+	ts.TxDualstakingClaimRewards(providerAAcc.Addr.String(), "")
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/rewards/keeper/providers_test.go:244]: done

File: x/rewards/keeper/providers_test.go [hunk 22]
```diff
@@ -0,0 +1,382 @@
...
+
+	// first months there are no bonus rewards, just payment ffrom the subscription
+	ts.AdvanceMonths(1)
+	ts.AdvanceBlocks(ts.BlocksToSave() + 1)
+
+	res, err := ts.QueryDualstakingDelegatorRewards(providerAAcc.Addr.String(), providerAAcc.Addr.String(), "")
+	require.Nil(t, err)
+	require.Len(t, res.Rewards, 1)
+	require.Equal(t, res.Rewards[0].Amount.Amount, ts.plan.Price.Amount.MulRaw(50))
+	ts.TxDualstakingClaimRewards(providerAAcc.Addr.String(), providerAAcc.Addr.String())
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/rewards/keeper/providers_test.go:191]: done

File: x/rewards/keeper/providers_test.go [hunk 23]
```diff
@@ -0,0 +1,382 @@
...
+	ts.TxDualstakingClaimRewards(providerAAcc.Addr.String(), "")
+
+	// now the provider should get all of the provider allocation
+	ts.AdvanceMonths(1)
+	ts.AdvanceEpoch()
+
+	res, err = ts.QueryDualstakingDelegatorRewards(providerAAcc.Addr.String(), providerAAcc.Addr.String(), "")
+	require.Nil(t, err)
+	require.Len(t, res.Rewards, 2)
+	require.Equal(t, res.Rewards[0].Amount.Amount, res.Rewards[1].Amount.Amount.MulRaw(2))
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/rewards/keeper/providers_test.go:308]: this is tested in the other tests, this test is to see that the shares are taken into consideration

File: x/rewards/keeper/providers_test.go [hunk 24]
```diff
@@ -0,0 +1,382 @@
...
+	ts.TxDualstakingClaimRewards(providerAAcc.Addr.String(), "")
+
+	// now the provider should get all of the provider allocation
+	ts.AdvanceMonths(1)
+	distBalance := ts.Keepers.Rewards.TotalPoolTokens(ts.Ctx, types.ProviderDistributionPool)
+	ts.AdvanceEpoch()
+
+	res, err = ts.QueryDualstakingDelegatorRewards(providerAAcc.Addr.String(), providerAAcc.Addr.String(), "")
+	require.Nil(t, err)
+	require.Len(t, res.Rewards, 1)
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/rewards/keeper/providers_test.go:253]: done

File: x/rewards/keeper/providers_test.go [hunk 25]
```diff
@@ -0,0 +1,382 @@
...
+	require.Equal(t, res.Rewards[0].Amount.Amount, distBalance)
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/rewards/keeper/providers_test.go:254]: done

File: x/rewards/keeper/providers_test.go [hunk 26]
```diff
@@ -0,0 +1,382 @@
+package keeper_test
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/rewards/keeper/providers_test.go:1]: done

File: x/rewards/keeper/providers.go [hunk 27]
```diff
@@ -0,0 +1,106 @@
+package keeper
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/rewards/keeper/providers.go:1]: done

File: x/rewards/keeper/providers.go [hunk 28]
```diff
@@ -0,0 +1,123 @@
...
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/lavanet/lava/utils"
+	"github.com/lavanet/lava/x/rewards/types"
+	subscriptionTypes "github.com/lavanet/lava/x/subscription/types"
+)
+
+func (k Keeper) AggregateRewards(ctx sdk.Context, provider, chainid string, adjustmentDenom uint64, rewards math.Int) {
+	index := types.BasePayIndex{Provider: provider, ChainID: chainid}
+	basepay, found := k.getBasePay(ctx, index)
+	adjustedPay := sdk.NewDecFromInt(rewards).QuoInt64(int64(adjustmentDenom))
```
Comment (omerlavanet) [inline x/rewards/keeper/providers.go:14]: let's do a redundant check here that adjustedPay is smaller eq than rewards

File: x/rewards/keeper/providers.go [hunk 29]
```diff
@@ -0,0 +1,123 @@
...
+			_, totalSent, err := k.dualstakingKeeper.RewardProvidersAndDelegators(ctx, providerAddr, basepay.ChainID, reward, string(types.ProviderRewardsDistributionPool), false, false, false)
+			if err != nil {
+				utils.LavaFormatError("Failed to send bonus rewards to provider", err, utils.LogAttr("provider", basepay.Provider))
+			}
+			totalToSendRewarded = totalToSendRewarded.Add(totalSent)
+		}
+	}
+	k.removeAllBasePay(ctx)
+	if !totalToSendRewarded.IsZero() {
+		err := k.bankKeeper.SendCoinsFromModuleToModule(ctx, string(types.ProviderRewardsDistributionPool), subscriptionTypes.ModuleName, sdk.NewCoins(sdk.NewCoin(k.stakingKeeper.BondDenom(ctx), totalToSendRewarded)))
```
Comment (omerlavanet) [inline x/rewards/keeper/providers.go:60]: we don't need this if the change in RewardProvidersAndDelegators sends the tokens while rewarding

File: x/rewards/keeper/providers.go [hunk 30]
```diff
@@ -0,0 +1,123 @@
...
+		if err != nil {
+			utils.LavaFormatError("failed to send bonus rewards to subscription module", err, utils.LogAttr("amount", totalRewarded.String()))
+		}
+	}
+
+	tokensToBurn := total.Sub(totalRewarded)
+	err := k.BurnPoolTokens(ctx, types.ProviderRewardsDistributionPool, tokensToBurn)
```
Comment (omerlavanet) [inline x/rewards/keeper/providers.go:67]: the burn will be in the same place validators are being burned

File: x/rewards/keeper/providers_test.go [hunk 31]
```diff
@@ -0,0 +1,420 @@
...
+	require.Nil(t, err)
+
+	res3, err := ts.QueryDualstakingDelegatorRewards(providerAcc3.Addr.String(), "", "")
+	require.Nil(t, err)
+	require.Len(t, res.Rewards, 1)
+	// we add 6 because of truncating
+	require.Equal(t, res3.Rewards[0].Amount.Amount, distBalance.QuoRaw(7).MulRaw(4).AddRaw(6))
+	_, err = ts.TxDualstakingClaimRewards(providerAcc3.Addr.String(), providerAcc3.Addr.String())
+	require.Nil(t, err)
+}
```
Comment (omerlavanet) [inline x/rewards/keeper/providers_test.go:420]: can we do a test that passes 49 months? so the boos runs out i want to see this edge case doesn't panic

**Review by omerlavanet — CHANGES_REQUESTED** [review]

File: x/rewards/keeper/providers.go (hunk 30 above)
Comment (Yaroms) [reply x/rewards/keeper/providers.go:67]: done

File: x/rewards/keeper/providers.go [hunk 32]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+func (k Keeper) specEmissionParts(ctx sdk.Context) (emisions []types.SpecEmmisionPart) {
+	chainIDs := k.specKeeper.GetAllChainIDs(ctx)
+	totalStake := sdk.ZeroDec()
+	chainStake := map[string]sdk.Dec{}
+	for _, chainID := range chainIDs {
+		spec, found := k.specKeeper.GetSpec(ctx, chainID)
+		if !found {
+			continue
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/rewards/keeper/providers.go:87]: done

File: x/rewards/keeper/providers.go (hunk 28 above)
Comment (Yaroms) [reply x/rewards/keeper/providers.go:14]: done

File: x/rewards/keeper/providers.go [hunk 33]
```diff
@@ -0,0 +1,123 @@
...
+		if stake, ok := chainStake[chainID]; ok {
+			if stake.IsZero() || totalStake.IsZero() {
+				continue
+			}
+
+			emisions = append(emisions, types.SpecEmmisionPart{ChainID: chainID, Emission: stake.Quo(totalStake)})
+		}
+	}
+
+	return emisions
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/rewards/keeper/providers.go:113]: done

File: x/rewards/keeper/providers.go [hunk 34]
```diff
@@ -0,0 +1,123 @@
...
+			chainStake[chainID] = chainStake[chainID].Add(sdk.NewDecFromInt(entry.EffectiveStake()))
+		}
+
+		chainStake[chainID] = chainStake[chainID].MulInt64(int64(spec.Shares))
+		totalStake = totalStake.Add(chainStake[chainID])
+	}
+
+	for _, chainID := range chainIDs {
+		if stake, ok := chainStake[chainID]; ok {
+			if stake.IsZero() || totalStake.IsZero() {
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/rewards/keeper/providers.go:105]: done

File: x/rewards/types/spec_emmision.go [hunk 35]
```diff
@@ -0,0 +1,8 @@
+package types
+
+import sdk "github.com/cosmos/cosmos-sdk/types"
+
+type SpecEmmisionPart struct {
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/rewards/types/spec_emmision.go:5]: done

File: x/rewards/keeper/rewards.go [hunk 36]
```diff
@@ -66,6 +66,7 @@ func (k Keeper) RefillRewardsPools(ctx sdk.Context, _ []byte, data []byte) {
 	}
 
 	k.refillDistributionPool(ctx, monthsLeft, types.ValidatorsRewardsAllocationPoolName, types.ValidatorsRewardsDistributionPoolName)
+	k.refillDistributionPool(ctx, monthsLeft, types.ProvidersRewardsAllocationPool, types.ProviderRewardsDistributionPool)
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/rewards/keeper/rewards.go]: done

File: x/rewards/keeper/providers.go [hunk 37]
```diff
@@ -0,0 +1,123 @@
...
+	}
+
+	tokensToBurn := total.Sub(totalRewarded)
+	err := k.BurnPoolTokens(ctx, types.ProviderRewardsDistributionPool, tokensToBurn)
+	if err != nil {
+		utils.LavaFormatError("Failed to burn left over bonus rewards", err, utils.LogAttr("amount", tokensToBurn.String()))
+	}
+}
+
+func (k Keeper) specTotalPayout(ctx sdk.Context, totalMonthlyPayout math.Int, totalProvidersBaseRewards sdk.Dec, spec types.SpecEmmisionPart) math.LegacyDec {
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/rewards/keeper/providers.go:68]: done

File: x/rewards/keeper/providers.go [hunk 38]
```diff
@@ -0,0 +1,123 @@
...
+		basepay = types.BasePay{Total: rewards, TotalAdjusted: adjustedPay}
+	} else {
+		basepay.Total = basepay.Total.Add(rewards)
+		basepay.TotalAdjusted = basepay.TotalAdjusted.Add(adjustedPay)
+	}
+
+	k.setBasePay(ctx, index, basepay)
+}
+
+func (k Keeper) distributeMonthlyBonusRewards(ctx sdk.Context) {
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/rewards/keeper/providers.go:26]: done

File: x/rewards/keeper/providers.go [hunk 39]
```diff
@@ -0,0 +1,123 @@
...
+	totalRewarded := sdk.ZeroInt()
+	totalToSendRewarded := sdk.ZeroInt()
+	specs := k.specEmissionParts(ctx)
+	for _, spec := range specs {
+		basepays, totalbasepay := k.specProvidersBasePay(ctx, spec.ChainID)
+		if totalbasepay.IsZero() {
+			continue
+		}
+
+		specTotalPayout := k.specTotalPayout(ctx, total, sdk.NewDecFromInt(totalbasepay), spec)
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/rewards/keeper/providers.go:37]: done

File: x/rewards/keeper/providers.go (hunk 29 above)
Comment (Yaroms) [reply x/rewards/keeper/providers.go:61]: done

File: x/dualstaking/keeper/delegator_reward.go [hunk 40]
```diff
@@ -156,22 +156,22 @@ func (k Keeper) ClaimRewards(ctx sdk.Context, delegator string, provider string)
 
 // RewardProvidersAndDelegators is the main function handling provider rewards with delegations
 // it returns the provider reward amount and updates the delegatorReward map with the reward portion for each delegator
-func (k Keeper) RewardProvidersAndDelegators(ctx sdk.Context, providerAddr sdk.AccAddress, chainID string, totalReward math.Int, senderModule string, calcOnlyProvider bool, calcOnlyDelegators bool, calcOnlyContributer bool) (providerReward math.Int, err error) {
+func (k Keeper) RewardProvidersAndDelegators(ctx sdk.Context, providerAddr sdk.AccAddress, chainID string, totalReward math.Int, senderModule string, calcOnlyProvider bool, calcOnlyDelegators bool, calcOnlyContributer bool) (providerReward math.Int, totalRewards math.Int, err error) {
 	block := uint64(ctx.BlockHeight())
 	epoch, _, err := k.epochstorageKeeper.GetEpochStartForBlock(ctx, block)
 	if err != nil {
-		return math.ZeroInt(), utils.LavaFormatError(types.ErrCalculatingProviderReward.Error(), err,
+		return math.ZeroInt(), math.ZeroInt(), utils.LavaFormatError(types.ErrCalculatingProviderReward.Error(), err,
```
[the remark this answers is not in the document]
Comment (Yaroms) [reply x/dualstaking/keeper/delegator_reward.go:163]: done

**Review by oren-lava — APPROVED** [review]
